Introduce layout constraints and requirements.

swiftix opened the pull request:
This PR introduces new kind of requirements: layout requirements.

This kind of requirements allows to expose that a type should satisfy certain layout properties, e.g. it should be a trivial type, have a given size and alignment, etc.

This is an NFC for now, because nothing produces these constraints yet. The functionality like parsing, code generation, etc will be submitted in subsequent PRs.

swiftix commented:
This is a self-contained part of the https://github.com/apple/swift/pull/6486

swiftix commented:
@slavapestov This is the first chunk of the partial specialization PR. It contains only data types for layout constraints and requirements as well as all the basic support for them. Could you please review?

slavapestov commented on include/swift/AST/Requirement.h:
```diff
@@ -43,8 +46,11 @@ enum class RequirementKind : unsigned {
-  llvm::PointerIntPair<Type, 2, RequirementKind> FirstTypeAndKind;
-  Type SecondType;
+  llvm::PointerIntPair<Type, 3, RequirementKind> FirstTypeAndKind;
+  union {
```
Should this be a PointerUnion?

swiftix replied:
I don't have a strong opinion. We can do it now, but should we add more alternatives in the future and have e.g. 6 different kinds of a payload, we probably don't want to use a PointerUnion for that, or? Plus, each getter already asserts if the tag is wrong.

swiftix replied:
And if one of those new alternatives will not be a pointer-like something, we won't be able to use union anymore. 

So, do you still think we should go for a PointerUnion?

slavapestov replied:
It's fine to leave it as-is, if you add a comment explaining that it must match the tag.

slavapestov commented on include/swift/AST/Types.h:
```diff
@@ -3283,8 +3290,11 @@ class SILFunctionType final : public TypeBase, public llvm::FoldingSetNode,
-  bool isPolymorphic() const { return GenericSig != nullptr; }
+
+  bool isPolymorphic() const {
+    return GenericSig != nullptr && !GenericSig->areAllParamsConcrete();
```
This should not be part of this patch -- we can discuss it later

swiftix replied:
Oops. It sneaked in. I'll remove it.

slavapestov replied:
Thanks!

swiftix commented:
@eeckstein Erik, could you have a look at this specific commit: https://github.com/apple/swift/pull/6743/commits/cbccb6a9fbd0f8fca8cfe06ff91625f6690c3043 ? It is the (de)mangling implementation for layout requirements.

slavapestov commented on lib/AST/LayoutConstraint.cpp:
```diff
@@ -0,0 +1,61 @@
+namespace swift {
+
+LayoutConstraintInfo getLayoutConstraintInfo(StringRef ID) {
+  if (ID == "_Trivial")
```
Should these be uniqued Identifiers in the ASTContext?

swiftix replied:
May be. Is it really worth it?

slavapestov replied:
I assume the identifiers will appear in a few other places (diagnostics, parser) and it's easier to keep everything in sync when/if they change in the future.

swiftix replied:
I see. But these identifiers are also used in some contexts, where ASTContext is unavailable (e.g. demangler, etc). What do we do about that?

swiftix replied:
May be instead I should define these identifiers by means of x-macros as static members of LayoutConstraint, so that it is possible to access them anywhere using`LayoutConstraint::Id_Trivial`, etc?

slavapestov replied:
Sure -- and you can use the x-macros to make the Identifiers too

slavapestov commented on lib/AST/Type.cpp:
```diff
@@ -1590,6 +1597,11 @@ bool TypeBase::isSpelledLike(Type other) {
 
+LayoutConstraint TypeBase::getLayoutConstraint() {
+  auto CanTy = getCanonicalType();
+  return CanTy.getLayoutConstraint();
```
There's no need to have the CanType method for this, just call getAs instead of dyn_cast

swiftix replied:
Sure.

slavapestov commented on include/swift/AST/ArchetypeBuilder.h:
```diff
@@ -129,12 +132,6 @@ class ArchetypeBuilder {
   ArchetypeBuilder(const ArchetypeBuilder &) = delete;
   ArchetypeBuilder &operator=(const ArchetypeBuilder &) = delete;
 
-  /// \brief Add a new conformance requirement specifying that the given
```
I don't think this change is necessary -- do these methods have to be public?

eeckstein commented on docs/ABI.rst:
```diff
@@ -1160,11 +1160,26 @@ Property behaviors are implemented using private protocol conformances.
+  LAYOUT-CONSTRAINT ::= 'U'  // Unknown layout
+
+  SIZE-AND-ALIGNMENT ::= INDEX // Size only
+  SIZE-AND-ALIGNMENT ::= INDEX INDEX // Size followed by alignment
```
This does not work, because it's not clear where the 'RM' operator ends. The second INDEX could be a second index or a following identifier (which starts with the length).

I suggest using different LAYOUT-CONSTRAINT letters, e.g. 'E' + 'e' and 'M' + 'm' to distinguish between a single and double INDEX.

swiftix replied:
Thanks! Good catch! I'll fix it.

eeckstein commented:
I just reviewed the _new_ mangling. LGTM, except the one comment I made.

swiftix commented:
@slavapestov @eeckstein I pushed the new version which is supposed to address all of the review comments.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
LGTM. Can you implement the serialization/deserialization for layout constraints as well? Otherwise serialized SIL partial specializations will not work in IRGen.

slavapestov commented:
@swift-ci Please smoke test

eeckstein commented on lib/Basic/Remangler.cpp:
```diff
@@ -679,6 +679,24 @@ void Remangler::mangleDependentGenericSameTypeRequirement(Node *node) {
+  assert(node->getChild(1)->getText().size() == 1);
+  Buffer << node->getChild(1)->getText()[0];
+  if (node->getNumChildren() >=3)
+    mangleChildNode(node, 2);
```
This look wrong. The children (which are the type parameters - I assume) should be mangled before the 'R...' operator. Also it seems that size, alignment mangling is missing.

swiftix replied:
The children (which represent the RHS of the requirement, i.e. the layout constraint itself) are: the letter for the kind of layout constraint and one or two numbers for size/alignment. No types. 

The only type in play is the LHS of the requirement. This is the dependent generic param index. So, IMHO, everything is OK.

slavapestov replied:
That's why tests would be good, @swiftix :)

swiftix replied:
Sure, working on this. I was not aware of test/Demangle/Inputs/manglings.txt until you mentioned it.

eeckstein replied:
ok, thanks for the explanation.

slavapestov commented:
@swiftix Can you add mangling tests to test/Demangle/Inputs/manglings.txt?

slavapestov commented:
Also, @eeckstein can we have a version of that test for the new mangling?

swiftix commented:
I've added mangling tests for the old mangling and fixed a bug ;-)
Do we have any tests for the new mangling?

Will look into serialization/deserialization now.

swiftix commented:
@swift-ci Please smoke test

swiftix commented:
@swift-ci Please smoke test

swiftix commented:
@eeckstein I added tests as was requested. Found 2 small uses along the way.

@slavapestov Serialization is implemented, but not tested yet as this patch does not parse and produce the layout constraints yet. This will be the next PR.

eeckstein commented:
> Also, @eeckstein can we have a version of that test for the new mangling?

@slavapestov I'll convert manglings.txt to the new mangling soon (next week).

swiftix commented:
I think I'll take serialization/deserialization out of this PR and include it into the subsequent one, so that we can test it there.

swiftix commented:
@swift-ci Please smoke test and merge

swiftix commented:
@swift-ci Please smoke test and merge
